src/Dom/Document.php

```diff
@@ -0,0 +1,968 @@
+		);
+
+		$node->parentNode->removeChild( $comment_start );
+		$node->parentNode->removeChild( $comment_end );
+
+		return $html;
+	}
+
+	/**
+	 * Normalize the document structure.
+	 *
+	 * This makes sure the document adheres to the general structure that AMP requires:
+	 *   ```
+	 *   <!doctype html>
+	 *   <html>
+	 *     <head>
+	 *       <meta charset="utf-8">
+	 *     </head>
+	 *     <body>
+	 *     </body>
+	 *   </html>
+	 *   ```
+	 *
+	 * @param string $content Content to normalize the structure of.
+	 * @return string Normalized content.
+	 */
+	private function normalize_document_structure( $content ) {
+		$matches    = [];
+		$html_start = '<html>';
+
+		// Strip <!doctype> for now.
+		$content = preg_replace( self::HTML_STRUCTURE_DOCTYPE_PATTERN, '', $content, 1 );
+
+		// Detect and strip <html> tags.
+		if ( preg_match( self::HTML_STRUCTURE_HTML_START_TAG, $content, $matches ) ) {
+			$html_start = $matches['html_start'];
+			$content    = preg_replace( self::HTML_STRUCTURE_HTML_START_TAG, '', $content, 1 );
+			$content    = preg_replace( self::HTML_STRUCTURE_HTML_END_TAG, '', $content, 1 );
+		}
+
+		// Detect <head> and <body> tags and add as needed.
+		if ( ! preg_match( self::HTML_STRUCTURE_HEAD_START_TAG, $content, $matches ) ) {
+			if ( ! preg_match( self::HTML_STRUCTURE_BODY_START_TAG, $content, $matches ) ) {
+				// Both <head> and <body> missing.
+				$content = "<head></head><body>{$content}</body>";
+			} else {
+				// Only <head> missing.
+				$content = "<head></head>{$content}";
+			}
+		} else {
+			if ( ! preg_match( self::HTML_STRUCTURE_BODY_END_TAG, $content, $matches ) ) {
+				// Only <body> missing.
+				// @todo This is an expensive regex operation, look into further optimization.
+				$content  = preg_replace( self::HTML_STRUCTURE_HEAD_TAG, '$0<body>', $content, 1 );
+				$content .= '</body>';
+			}
+		}
+
+		$content = "{$html_start}{$content}</html>";
+
+		// Reinsert a standard doctype.
+		$content = "<!DOCTYPE html>{$content}";
+
+		return $content;
+	}
+
+	/**
+	 * Normalize the structure of the document if it was already provided as a DOM.
+	 */
+	public function normalize_dom_structure() {
+		$head = $this->getElementsByTagName( self::TAG_HEAD )->item( 0 );
+		if ( ! $head ) {
+			$this->head = $this->createElement( self::TAG_HEAD );
+			$this->insertBefore( $this->head, $this->firstChild );
+		}
+
+		$body = $this->getElementsByTagName( self::TAG_BODY )->item( 0 );
+		if ( ! $body ) {
+			$this->body = $this->createElement( self::TAG_BODY );
+			$this->appendChild( $this->body );
+		}
+
+		$this->move_invalid_head_nodes_to_body();
+	}
+
+	/**
+	 * Move invalid head nodes back to the body.
+	 */
+	private function move_invalid_head_nodes_to_body() {
+		// Walking backwards makes it easier to move elements in the expected order.
+		$node = $this->head->lastChild;
+		while ( $node ) {
+			$next_sibling = $node->previousSibling;
+			if ( ! AMP_DOM_Utils::is_valid_head_node( $node ) ) {
```

Comment:
It is in the Document class...?

Reply:
Oh, nevermind, looked at the wrong method in the file view.

The `AMP_DOM_Utils::is_valid_head_node()` is publicly accessible, so moving it would mean deprecating the old version, not removing that one. But I can do that as well.

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/3758/commits/41094b488fa7977effd9e25e1473f7197169ea6b